### PR TITLE
Invariant Risk Minimization

### PR DESCRIPTION
**Pull request by ben-heil**
Changes:
- Implement invariant risk minimization for training classifiers
- Switch both the vanilla batch gradient descent and irm to use BCEWithLogits loss (the other BCE loss function in pytorch doesn't have as second derivative implemented for some reason
- Update dataset.py and model.py to be compatible with BCEWithLogits
- Add result saving functionality to classifier.py and util.py

### CODE REVIEW

File: whistl/classifier.py [hunk 1]
```diff
@@ -3,106 +3,212 @@
...
+    tune_label_counts, _ = util.get_value_counts(tune_loader)
+    baseline = max(list(tune_label_counts.values())) / len(tune_dataset)
 
+    results = {'train_loss': [], 'tune_loss': [], 'train_acc': [], 'tune_acc': [],
+               'baseline': baseline, 'train_penalty': [], 'train_raw_loss': []}
+
+    try:
+        # TODO there is probably a more concise way of handling this
+        train_samples = 0
+        for study_count in train_study_counts:
```
Comment (jjc2718) [inline whistl/classifier.py]: If you want, you can replace this with
```python
train_samples = sum(train_study_counts)
```
(probably still just a for loop under the hood, but slightly more concise)

File: whistl/classifier.py [hunk 2]
```diff
@@ -3,106 +3,212 @@
...
+            train_samples += study_count
+
+        dummy_w = torch.nn.Parameter(torch.DoubleTensor([1.0]))
+        # Create dictionary of parameters for visualizing operations
+        params = dict(classifier.named_parameters())
+        params['dummy_w'] = dummy_w
```
Comment (jjc2718) [inline whistl/classifier.py]: It doesn't look like you're doing anything with this dictionary currently - is that intentional?

File: whistl/classifier.py [hunk 3]
```diff
@@ -3,106 +3,212 @@
...
-    # Performance impact should be negligible though
-    correct = 0
-    predictions = [1 if p > .5 else 0 for p in output]
-    for y, y_hat in zip(predictions, labels):
-        if y == y_hat:
-            correct += 1
-    return correct
+def compute_irm_penalty(loss, dummy_w):
+    ''' Calculate the invariance penalty for the classifier'''
+    dummy_grad = abs(grad(loss, dummy_w, create_graph=True)[0])
```
Comment (jjc2718) [inline whistl/classifier.py:27]: Might be nice to have a short high-level explanation of what this is doing here. I know you linked to the paper, but it takes a bit of time/effort to read and understand.

Something like "Calculate the invariance penalty for the classifier. This is defined in the paper as the norm of the gradient of the loss function for an input of 1." (or something else if what I wrote doesn't make sense, I'm still not totally sure I understand it)

You could even write out the formula from the paper in some sort of pseudo-Latex (I do this sometimes: https://github.com/greenelab/netreg/blob/master/data_models.py#L405) if you think that makes things clearer.

File: whistl/classifier.py [hunk 4]
```diff
@@ -3,106 +3,212 @@
...
 
-def parse_map_file(map_file_path):
-    '''Create a sample: label mapping from the pickled file output by label_samples.py
+
+def train_with_irm(classifier, map_file, train_dirs, tune_dirs, gene_file,
+                   num_epochs, loss_scaling_factor, label_to_encoding, device, logger=None):
+    '''Train the provided classifier using invariant risk minimization
+    IRM looks for features in the data that are invariant between different environments, as
+    they are more likely to be predictive of true causal signals as opposed to spurious
+    correlations. For more information, read https://arxiv.org/abs/1907.02893
```
Comment (jjc2718) [inline whistl/classifier.py:37]: General conceptual question: what do you expect IRM to give you in the context of your project? Do you think it will just give you better predictions, or does it have an interpretability advantage as well (i.e. some sort of measurement of which features are invariant/possibly causal)? If the latter, is there a simple way to see which features are invariant?

**Review by jjc2718 — APPROVED** [review]
The code looks good! My outstanding questions are mostly conceptual (but I assume you've been discussing similar questions in your meetings with Casey, so feel free to ignore if you already have a good handle on these things).

Other than what I wrote in the comments, the other thing I'm curious about is how you plan to validate that this is working (particularly if you're planning on using this for interpretation: how do you know this is actually finding causal features?). I'm guessing that's coming in future PRs, but worth thinking about if you're not already.

File: whistl/classifier.py (hunk 2 above)
Comment (ben-heil) [reply whistl/classifier.py]: Good catch, I was using it to find which operation wasn't implemented by visualizing with torchviz. I'll remove it since I'm not using it anymore

File: whistl/classifier.py (hunk 4 above)
Comment (ben-heil) [reply whistl/classifier.py:37]: The authors claim that the data representation you get from training a model with IRM is more able to learn which features truly matter, by training your model to perform well in different environments. The questions I'm currently asking are
1) Does IRM work with a large number of environments (like the number of studies related to a disease
2) Does IRM give a generalization boost by helping the model ignore batch effects
